File: CS/WebApplication1/CustomDataSource.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Web;

namespace WebApplication1 {
    public class CustomAppointmentDataSource {
        BindingList<CustomAppointment> appts;
        public CustomAppointmentDataSource(BindingList<CustomAppointment> parAppts) {
            if(parAppts == null)
                DevExpress.XtraScheduler.Native.Exceptions.ThrowArgumentNullException("Appointments");
            this.Appointments = parAppts;
        }
        public CustomAppointmentDataSource()
            : this(new BindingList<CustomAppointment>()) {

        }
        public BindingList<CustomAppointment> Appointments { 
            get { return appts; } 
            set {
                if(value is BindingList<CustomAppointment> && (value as BindingList<CustomAppointment>).Count > 0 ) {
                    appts = value;
                    for(int i = 0; i < appts.Count; i++) {
                        appts[i].Id = i;
                    }
                }
            } 
        }

        #region ObjectDataSource methods                        
        public CustomAppointment GetSourceAppointment(object apptId) {
            foreach(CustomAppointment item in Appointments) {
                if(Convert.ToInt32(item.Id) == Convert.ToInt32(apptId)) {
                    return item;
                }
            }
            return null;
        }

        public object InsertMethodHandler(CustomAppointment customAppt) {
            int lastId = 0;
            foreach (CustomAppointment item in Appointments)
	        {
                if (Convert.ToInt32(item.Id) > lastId) 
                    lastId = Convert.ToInt32(item.Id);		 
	        }
            customAppt.Id = lastId + 1;
            Appointments[Appointments.Count - 1].Id = customAppt.Id;
            return customAppt.Id;
        }
        public void DeleteMethodHandler(CustomAppointment customAppt) {
            CustomAppointment currentAppt = GetSourceAppointment(customAppt.Id);
            if(customAppt != null)
                Appointments.Remove(customAppt);
        }
        public void UpdateMethodHandler(CustomAppointment customAppt) {
            CustomAppointment currentAppt = GetSourceAppointment(customAppt.Id);
            if(customAppt != null) {
                currentAppt.AllDay = customAppt.AllDay;
                currentAppt.Description = customAppt.Description;
                currentAppt.EndTime = customAppt.EndTime;
                currentAppt.EventType = customAppt.EventType;
                currentAppt.Id = customAppt.Id;
                currentAppt.Label = customAppt.Label;
                currentAppt.Location = customAppt.Location;
                currentAppt.OwnerId = customAppt.OwnerId;
                currentAppt.RecurrenceInfo = customAppt.RecurrenceInfo;
                currentAppt.ReminderInfo = customAppt.ReminderInfo;
                currentAppt.StartTime = customAppt.StartTime;
                currentAppt.Status = customAppt.Status;
                currentAppt.Subject = customAppt.Subject;
            }
        }
        public IEnumerable SelectMethodHandler() {
            return Appointments;
        }
        #endregion
    }

    public class CustomResourceDataSource {
        BindingList<CustomResource> resources;
        public CustomResourceDataSource(BindingList<CustomResource> parResources) {
            if(parResources == null)
                DevExpress.XtraScheduler.Native.Exceptions.ThrowArgumentNullException("Resources");
            this.Resources = parResources;
        }
        public CustomResourceDataSource() : this(new BindingList<CustomResource>()) {

        }
        public BindingList<CustomResource> Resources { 
            get { return resources; } 
            set { resources = value; } 
        }

        #region ObjectDataSource methods
        public IEnumerable SelectMethodHandler() {
            return Resources;
        }
        #endregion
    }
}

File: VB/WebApplication1/CustomDataSource.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Linq
Imports System.Web

Namespace WebApplication1
	Public Class CustomAppointmentDataSource
		Private appts As BindingList(Of CustomAppointment)
		Public Sub New(ByVal parAppts As BindingList(Of CustomAppointment))
			If parAppts Is Nothing Then
				DevExpress.XtraScheduler.Native.Exceptions.ThrowArgumentNullException("Appointments")
			End If
			Me.Appointments = parAppts
		End Sub
		Public Sub New()
			Me.New(New BindingList(Of CustomAppointment)())

		End Sub
		Public Property Appointments() As BindingList(Of CustomAppointment)
			Get
				Return appts
			End Get
			Set(ByVal value As BindingList(Of CustomAppointment))
				If TypeOf value Is BindingList(Of CustomAppointment) AndAlso (TryCast(value, BindingList(Of CustomAppointment))).Count > 0 Then
					appts = value
					For i As Integer = 0 To appts.Count - 1
						appts(i).Id = i
					Next i
				End If
			End Set
		End Property

		#Region "ObjectDataSource methods                        "
		Public Function GetSourceAppointment(ByVal apptId As Object) As CustomAppointment
			For Each item As CustomAppointment In Appointments
				If Convert.ToInt32(item.Id) = Convert.ToInt32(apptId) Then
					Return item
				End If
			Next item
			Return Nothing
		End Function

		Public Function InsertMethodHandler(ByVal customAppt As CustomAppointment) As Object
			Dim lastId As Integer = 0
			For Each item As CustomAppointment In Appointments
				If Convert.ToInt32(item.Id) > lastId Then
					lastId = Convert.ToInt32(item.Id)
				End If
			Next item
			customAppt.Id = lastId + 1
			Appointments(Appointments.Count - 1).Id = customAppt.Id
			Return customAppt.Id
		End Function
		Public Sub DeleteMethodHandler(ByVal customAppt As CustomAppointment)
			Dim currentAppt As CustomAppointment = GetSourceAppointment(customAppt.Id)
			If customAppt IsNot Nothing Then
				Appointments.Remove(customAppt)
			End If
		End Sub
		Public Sub UpdateMethodHandler(ByVal customAppt As CustomAppointment)
			Dim currentAppt As CustomAppointment = GetSourceAppointment(customAppt.Id)
			If customAppt IsNot Nothing Then
				currentAppt.AllDay = customAppt.AllDay
				currentAppt.Description = customAppt.Description
				currentAppt.EndTime = customAppt.EndTime
				currentAppt.EventType = customAppt.EventType
				currentAppt.Id = customAppt.Id
				currentAppt.Label = customAppt.Label
				currentAppt.Location = customAppt.Location
				currentAppt.OwnerId = customAppt.OwnerId
				currentAppt.RecurrenceInfo = customAppt.RecurrenceInfo
				currentAppt.ReminderInfo = customAppt.ReminderInfo
				currentAppt.StartTime = customAppt.StartTime
				currentAppt.Status = customAppt.Status
				currentAppt.Subject = customAppt.Subject
			End If
		End Sub
		Public Function SelectMethodHandler() As IEnumerable
			Return Appointments
		End Function
		#End Region
	End Class

	Public Class CustomResourceDataSource
		Private resources_Renamed As BindingList(Of CustomResource)
		Public Sub New(ByVal parResources As BindingList(Of CustomResource))
			If parResources Is Nothing Then
				DevExpress.XtraScheduler.Native.Exceptions.ThrowArgumentNullException("Resources")
			End If
			Me.Resources = parResources
		End Sub
		Public Sub New()
			Me.New(New BindingList(Of CustomResource)())

		End Sub
		Public Property Resources() As BindingList(Of CustomResource)
			Get
				Return resources_Renamed
			End Get
			Set(ByVal value As BindingList(Of CustomResource))
				resources_Renamed = value
			End Set
		End Property

		#Region "ObjectDataSource methods"
		Public Function SelectMethodHandler() As IEnumerable
			Return Resources
		End Function
		#End Region
	End Class
End Namespace